doc/design/api.md

```diff
@@ -0,0 +1,90 @@
+# Design Doc: PaddlePaddle API
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
```

Comment:
One comment for future when we will add bindings for more languages (or even now, if people use python to create topology and train with c++)

Now model parameter is initialized during model creation: `model = paddle.model.create()`.
`model.save()` will always save both `parameter` and `topology`, and `model.load()` will load both as well.
We may need a method `model.init()` to initialize model parameters again.

The use case is: When user create topology from python (because it's easier), and use it in C++ (or golang in future). User will like to have the option to init parameter again in order to start training from somewhere else in the parameter space.

From the use case above, we can see that topology and parameter is not tightly coupled. Maybe we can consider instead expose
```
save_topology()
save_parameter()
load_topology()
load_parameter()
```